### ISSUE TITLE
Exporting certificates created on the fly broken on OSX?

### ISSUE THREAD

**Issue by javiercn** [dotnet/corefx#20272]
We ran into this while trying to generate certificates on the fly to use for HTTPS. If you create a self-signed certificate from an ephemeral RSA key. Shouldn't you be able to export it to a PFX file directly?

it was failing for us in line:
```c#
var export = certificate.Export(X509ContentType.Pkcs12, "");
```

```c#
var signingRequest = new CertificateRequest(
    new X500DistinguishedName(TestSubject), rsa, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);

var enhancedKeyUsage = new OidCollection();
enhancedKeyUsage.Add(new Oid("1.3.6.1.5.5.7.3.1", "Server Authentication"));
signingRequest.CertificateExtensions.Add(new X509EnhancedKeyUsageExtension(enhancedKeyUsage, critical: true));
signingRequest.CertificateExtensions.Add(new X509KeyUsageExtension(X509KeyUsageFlags.KeyEncipherment, critical: true));
signingRequest.CertificateExtensions.Add(
    new X509BasicConstraintsExtension(
        certificateAuthority: false,
        hasPathLengthConstraint: false,
        pathLengthConstraint: 0,
        critical: true));

var sanBuilder = new SubjectAlternativeNameBuilder();
sanBuilder.AddDnsName(TestSubject.Replace("CN=", ""));
signingRequest.CertificateExtensions.Add(sanBuilder.Build());

var certificate = signingRequest.CreateSelfSigned(DateTimeOffset.UtcNow, DateTimeOffset.UtcNow.AddYears(1));

var export = certificate.Export(X509ContentType.Pkcs12, "");
var imported = new X509Certificate2(export, "", X509KeyStorageFlags.PersistKeySet);
Array.Clear(export, 0, export.Length);
```

**Comment by karelz** [comment]
@steveharter @ianhays can you please advise?
We discussed similar things with Jeremy, but I don't remember details.

Maybe this may be helpful: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/cross-platform-cryptography.md

**Comment by bartonjs** [comment]
This fell through my post-vacation cracks.

Looks like there's a lifetime management problem in CopyWithPrivateKey when a cert-from-not-in-a-keychain and a key-from-not-in-a-keychain get together.  When the cert gets added to the keychain to mate the SecIdentityRef internal cert->keychain pointer is set to the temporary keychain (and not cleared later).  When we dispose that cert we see it's part of a temporary keychain, downref the keychain, and delete it, before the PFX can be exported.  (It shouldn't have reported being part of that keychain, so it shouldn't have downref'd).

Let's see if there's still enough runway for 2.0...

**Comment by bartonjs** [comment]
Fixed in 2.0 and 2.1
